### ISSUE TITLE
Breaks TypeScript line endings with tsfmt

### ISSUE THREAD

**Issue by GrzegorzKozub** [sbdchd/neoformat#37]
Neoformat breaks `.ts` files line endings while Autoformat does not have this issue. This does not happen for `.js` files so could be formatter related. Running `tsfmt ` from console using `--stdin` does not break line endings.

![tsfmt](https://cloud.githubusercontent.com/assets/154056/21883359/0017a0fa-d8af-11e6-8698-31527b505210.gif)

Using neoformat 4d1f13d, typescript-formatter 4.0.1 and typescript 2.1.4 on Windows.

**Comment by sbdchd** [comment]
Hmm. This [issue][1] has come up before with `tsfmt` and line endings.

Can you apply the following patch and copy the output of `:messages` after running `:Neoformat`? Thanks!

```diff
diff --git a/autoload/neoformat.vim b/autoload/neoformat.vim
index 4290a74..a071f0d 100644
--- a/autoload/neoformat.vim
+++ b/autoload/neoformat.vim
@@ -69,8 +69,10 @@ function! s:neoformat(bang, user_input, start_line, end_line) abort
 
         call neoformat#utils#log(cmd.exe)
         if cmd.stdin
+            call neoformat#utils#log('using stdin')
             let stdout = systemlist(cmd.exe, stdin)
         else
+            call neoformat#utils#log('using tmp file')
             call mkdir('/tmp/neoformat', 'p')
             call writefile(stdin, cmd.tmp_file_path)
             let stdout = systemlist(cmd.exe)
```

[1]: https://github.com/sbdchd/neoformat/issues/15

**Comment by GrzegorzKozub** [comment]
Sure, here it goes:
```
"d:\Code\neoformat\input.ts" 6L, 132C
"input.ts" 6 lines --83%--
Neoformat: ['function greeter(person) {', 'return "Hello, " + person;', '}', '  let user = "Jane User";', '  document.body.innerHTML = greeter(user);', '']
Neoformat: tsfmt --stdin d:\Code\neoformat\input.ts
Neoformat: using stdin
Neoformat: ['function greeter(person) {^M', '    return "Hello, " + person;^M', '}^M', 'let user = "Jane User";^M', 'document.body.innerHTML = greeter(user);^M']
Neoformat: tsfmt formatted buffer
```

**Comment by GrzegorzKozub** [comment]
Seems `systemlist` is to blame as `stdin` var is ok and then `stdout` has the line endings broken.

**Comment by GrzegorzKozub** [comment]
Based on `:h system` and `:h systemlist` this fixed the `tsfmt` problem with `--stdin`:

```
diff --git a/autoload/neoformat.vim b/autoload/neoformat.vim
index 4290a74..ee14641 100644
--- a/autoload/neoformat.vim
+++ b/autoload/neoformat.vim
@@ -69,7 +69,7 @@ function! s:neoformat(bang, user_input, start_line, end_line) abort
 
         call neoformat#utils#log(cmd.exe)
         if cmd.stdin
-            let stdout = systemlist(cmd.exe, stdin)
+            let stdout = split(system(cmd.exe, stdin), '\n')
         else
             call mkdir('/tmp/neoformat', 'p')
             call writefile(stdin, cmd.tmp_file_path)
```

**Comment by sbdchd** [comment]
Can you try the latest dev branch and tell me if it fixes your problem?

Thanks!

**Comment by GrzegorzKozub** [comment]
Yes, `dev` branch does fix this problem indeed. Thanks!

**Comment by iammerrick** [comment]
Getting same issue with latest.

**Comment by sbdchd** [comment]
@iammerrick Which version of `tsfmt` are you running?

**Comment by iammerrick** [comment]
@sbdchd 5.1.3.

**Comment by sbdchd** [comment]
```
❯ tsfmt --version
4.0.1
```

Seems to work.

Does the following fix the issue?

```diff
diff --git a/autoload/neoformat/formatters/typescript.vim b/autoload/neoformat/formatters/typescript.vim
index 30bb3fa..296f598 100644
--- a/autoload/neoformat/formatters/typescript.vim
+++ b/autoload/neoformat/formatters/typescript.vim
@@ -5,7 +5,7 @@ endfunction
 function! neoformat#formatters#typescript#tsfmt() abort
     return {
         \ 'exe': 'tsfmt',
-        \ 'args': ['--stdin', '%:p'],
-        \ 'stdin': 1
+        \ 'args': ['--replace'],
+        \ 'replace': 1
         \ }
 endfunction
```

**Comment by iammerrick** [comment]
@sbdchd Yes! That totally fixes the issue!

**Comment by iammerrick** [comment]
Have you considered allowing arbitrary arguments for formatters? Or is the recommendation there just make your own custom formatter?

**Comment by sbdchd** [comment]
I am open to passing arguments to formatters, but I think it is easiest to just make a new formatter definition.

**Comment by mhartington** [comment]
Hi there! So by changing how the formatter works, it now will not follow the style rules in the tsmft.json file.

```
 $ tsfmt --version
tsfmt : 5.2.0
tsc   : 2.3.2
```


![neofmt](https://cloud.githubusercontent.com/assets/2835826/25971333/e0a5be46-3669-11e7-9533-4fe8ce6f9259.gif)


In the first half, the lines are indented by 4 spaces while my settings are configured for 2. 
Switching back to using `--stdin` seems to resolve this. 

Now sure if this could be windows specific?

**Comment by sbdchd** [comment]
@mhartington I see that you are using a newer version of `tsfmt`, version `5.2.0` vs `5.1.3`.

If the stdin works for the latest version then I think the formatter should just be switched back to use stdin.

**Comment by mhartington** [comment]
@GrzegorzKozub  could you verify that it is okay to move back to stdin on windows with the newer version?

**Comment by GrzegorzKozub** [comment]
@mhartington @sbdchd I just did some testing on Windows 10 1703 using `neoformat` from `master` at `a0a3426`, `tsfmt 5.2.0` and `tsc 2.3.2`.

I can confirm that `neoformat` does not follow the rules defined in `tsfmt.json` or `.editorconfig` with `--replace` whereas it does follow them with `--stdin`. Side note: using bare `tsfmt` with `--replace` from terminal does not exhibit this problem.

In either case I did not observe incorrect line endings anymore, including without any rules file present.

So I'm all `--stdin`.
